### ISSUE TITLE
Warn if JSON schema file is valid, but an "empty schema"

### ISSUE THREAD

**Issue by christian-weiss** [Julian/jsonschema#636]
A JSON schema file that contains just `{}` is a vaild JSON schema as by draft-03, draft-04, draft-06 and draft-07. Same is true for a JSON schema file that contains at "root level" no typical JSON schema elements, but custom element names only.
(see: https://tools.ietf.org/html/draft-zyp-json-schema-03#section-5 for details) 
These JSON schema files can be called "emtpy schema", as they are valid schema files but without any validation rule. 

In fact, if an empty schema file is used to validate a JSON instance, then this instance is always valid (as it has no chance to fail; as there is no validation rule defined in that schema). A good validator should not error on "empty schema" files by default, but it should print a warning/info that this schema file is an "empty schema". An empty schema is not helpful, as it generates no business value and may lead into a false feeling of having all JSON instances successful validated. 

To not introduce a backward compatibility break i recommend to implement a `--strict` flag.
When `--strict`is set, then the validator should fail + exit (with exit code != 0) when an "empty schema" is detected.

This is would be helpful for draft-03, draft-04, draft-06 and draft-07. At least on draft-07 there is  a small hint in the official JSON Schema documentation, but on all other drafts that is missing, which makes it hard for "first time users" (of JSON schema an validators) to spot that "empty schema" problem.

**Comment by Julian** [comment]
I'm not sure I follow/agree with your assertions.

An empty schema is useful in many of the same ways that say, the identity function is.

**Comment by christian-weiss** [comment]
JSON schema + JSON instances are agnostic to programming languages. I think you could use a JSON instance directly without an "empty schema" the same way as you can use it as with an "empty schema".

JSON data producers can be implemented in programming language A, validator in B and consumer in C. If someone wants to use an identity function in the prefered programming language then he should be able to to that style/flavor independent from the fact that there is a schema, no schema or an empty schema.

If your code do not make assumption on the data structure of the JSON data, then you don't need a schema to validate it (but you may still want to validate it to ensure structure is consistent between datasets). If there is no need to validate the data structure, then i do not see a point in having an "empty schema" instead of no schema at all.

If i do not get your point: please explain your argument more in detail.

**Comment by Julian** [comment]
You're assuming one single scenario where the one schema someone is using
is the empty schema.

There are many others. Consider for example where an instance is fixed, and
someone is providing which schema to validate it with.

It's just not an error or warning condition -- same as if you wouldn't warn
if you were writing add(x, y) and someone passed y=0.

Sometimes the null thing is what sometime needs for some broader situation,
and warnings are just noise for them.

**Comment by christian-weiss** [comment]
I am trying hard, but i do not get your point.

Could you give a a concrete example of when a "empty schema" adds a benefit compared to NO SCHEMA? A sample JSON instance data + minimalistic implementation that uses an "empty schema" on that instance/data, that demonstrates why having an "empty schema" is better then NO SCHEMA. If you want to use an "empty schema" that is not equal to "{}" please also provide that schema.

**Comment by Julian** [comment]
I apologize for not really having a ton of time to write this out.

But I'd again basically encourage you to think about your last comment in the context of simple addition, in which what you're saying is "why bother having 0, if you don't have any number of things just don't write anything down at all".

It just turns out sometimes you still do want to write something down.

It's not for this library to decide when or why that may be, the user should decide, in the same spirit as "we are all consenting adults".

There just isn't any upside to such a warning.

Hopefully some of the above now starts to make sense, I'm going to close this since it's just not going to happen, but if it still doesn't feel free to respond and I'll make some effort when I have more spare cycles to have the dialog.

**Comment by christian-weiss** [comment]
I kindly ask to implement a `--strict` option and only warn on "empty schema" when `--strict` flag is set.

**Comment by Julian** [comment]
With apologies, not going to happen -- I'm especially not in favor of arbitrary "strict" modes with no standard definition and no good way to have backwards compatibility.

If you want such a thing, you're free to create an alias or function that'd be fairly simple:

```sh
function jsonschema() {
    command jsonschema -i "${(P)#}" =(echo '{"not": {"enum": [{}]}}') && command jsonschema $@
}```

(not tested, but I'm sure you can see what's there).
